### PR TITLE
ci: build prqlc on PRs

### PR DESCRIPTION
**Pull request by eitsupi**
Addresses #1910

### CODE REVIEW

**Comment by eitsupi** [comment]
It seems working.
https://github.com/PRQL/prql/actions/runs/4971568371
image

File: .github/actions/build-prqlc/action.yaml
Comment (eitsupi) [inline .github/actions/build-prqlc/action.yaml]: This doesn't seem working.

**Comment by eitsupi** [comment]
After a struggle I gave up on building `aarch64-unknown-linux-musl` target on amd64 linux. Could it be a [psm](https://github.com/rust-lang/stacker/tree/master/psm#platform-support) problem?

```sh
$ cargo build -p prqlc --release --target=aarch64-unknown-linux-musl
   Compiling cfg-if v1.0.0
   Compiling once_cell v1.17.1
   Compiling bitflags v1.3.2
   Compiling linux-raw-sys v0.1.4
   Compiling libc v0.2.140
   Compiling lazy_static v1.4.0
   Compiling memchr v2.5.0
   Compiling concolor-override v1.0.0
   Compiling concolor-query v0.3.3
   Compiling utf8parse v0.2.1
   Compiling adler v1.0.2
   Compiling gimli v0.27.2
   Compiling anstyle v0.3.5
   Compiling serde v1.0.157
   Compiling psm v0.1.21
   Compiling pin-project-lite v0.2.9
error: failed to run custom build command for `psm v0.1.21`

Caused by:
  process didn't exit successfully: `/workspaces/prql/target/release/build/psm-bd89755e99a601de/build-script-build` (exit status: 1)
  --- stdout
  OPT_LEVEL = Some("3")
  TARGET = Some("aarch64-unknown-linux-musl")
  HOST = Some("x86_64-unknown-linux-gnu")
  cargo:rerun-if-env-changed=CC_aarch64-unknown-linux-musl
  CC_aarch64-unknown-linux-musl = None
  cargo:rerun-if-env-changed=CC_aarch64_unknown_linux_musl
  CC_aarch64_unknown_linux_musl = None
  cargo:rerun-if-env-changed=TARGET_CC
  TARGET_CC = None
  cargo:rerun-if-env-changed=CC
  CC = None
  RUSTC_LINKER = None
  cargo:rerun-if-env-changed=CROSS_COMPILE
  CROSS_COMPILE = None
  cargo:rerun-if-env-changed=CFLAGS_aarch64-unknown-linux-musl
  CFLAGS_aarch64-unknown-linux-musl = None
  cargo:rerun-if-env-changed=CFLAGS_aarch64_unknown_linux_musl
  CFLAGS_aarch64_unknown_linux_musl = None
  cargo:rerun-if-env-changed=TARGET_CFLAGS
  TARGET_CFLAGS = None
  cargo:rerun-if-env-changed=CFLAGS
  CFLAGS = None
  cargo:rerun-if-env-changed=CRATE_CC_NO_DEFAULTS
  CRATE_CC_NO_DEFAULTS = None
  DEBUG = Some("false")
  CARGO_CFG_TARGET_FEATURE = Some("neon")
  cargo:rustc-cfg=asm
  cargo:rustc-cfg=switchable_stack
  running: "aarch64-linux-musl-gcc" "-O3" "-ffunction-sections" "-fdata-sections" "-fPIC" "-Wall" "-Wextra" "-xassembler-with-cpp" "-DCFG_TARGET_OS_linux" "-DCFG_TARGET_ARCH_aarch64" "-DCFG_TARGET_ENV_musl" "-o" "/workspaces/prql/target/aarch64-unknown-linux-musl/release/build/psm-9cc78c5eb0f1a15f/out/src/arch/aarch_aapcs64.o" "-c" "src/arch/aarch_aapcs64.s"

  --- stderr


  error occurred: Failed to find tool. Is `aarch64-linux-musl-gcc` installed?
```

File: .github/actions/build-prqlc/action.yaml
Comment (max-sixty) [inline .github/actions/build-prqlc/action.yaml:18]: Could we add an entry in https://github.com/PRQL/prql/blob/2bb271e9baabc1529649d64434e3ee1768ae1e1a/prql-compiler/Cargo.toml#L114-L115 for this?

File: .github/actions/build-prqlc/action.yaml
Comment (max-sixty) [reply .github/actions/build-prqlc/action.yaml]: WDYM by it not working? 

If it's that downloading the file from GitHub doesn't retain these permissions — I would guess that's a constraint from the downloading rather than this step (or any download could be marked as executable).

File: .github/actions/build-prqlc/action.yaml
Comment (max-sixty) [inline .github/actions/build-prqlc/action.yaml]:
```suggestion
        path: target/${{ inputs.target }}/release/${{ runner.os == 'Windows' && 'prqlc.exe' || 'prqlc' }}
```

...is maybe less explicit but more concise, will mean we can remove most of the previous two steps (your call on which is overall better...)

File: .github/workflows/pull-request.yaml
Comment (max-sixty) [inline .github/workflows/pull-request.yaml:112]: Before merging, I think probably we have this in `test-all` rather than `pull-request`  — or are there times it would fail on new code that we'd want to know about?

File: .github/actions/build-prqlc/action.yaml
Comment (max-sixty) [inline .github/actions/build-prqlc/action.yaml]: Is there any chance that `-C link-arg=-lgcc` needs quotes around it? (I don't know this well so please discard if you do...)

**Review by max-sixty — APPROVED** [review]
Excellent!! Thanks a lot.

A couple of points:
- No stress about Linux arm64. I see some comments in https://github.com/rust-lang/rust/issues/46651#issuecomment-433611633, not sure if they would help, but fine to leave for another day. 
  - I've also heard about https://github.com/cross-rs/cross, but don't have any experience with it. To the extent this can encapsulate much of the complication around non-default targets, that could work well. I don't see obvious docs for how to use with GHA — not sure whether that's because it's so easy or so difficult! :)
- I'm not sure whether this necessarily needs to be an "action" rather than a normal workflow, especially now that GH has loosened the restrictions on workflows calling other workflows. But it's also fine as an action.

- Do you know what we'd have to do to get these publishing on each release? We may have to auto-release IIUC. That would be a nice use of this.

File: .github/actions/build-prqlc/action.yaml
Comment (eitsupi) [reply .github/actions/build-prqlc/action.yaml]: I think this is the correct way to specify it, since enclosing it in quotation marks failed.
Searching across GitHub, I can see other projects that do it this way.

File: .github/workflows/pull-request.yaml
Comment (eitsupi) [reply .github/workflows/pull-request.yaml:112]: My idea was to provide an easy way to try out features added by pull requests.
For example, I believe duckdb builds and uploads duckdb CLI executables with every pull request.

**Comment by eitsupi** [comment]
Thank you for reviewing this.

I considered cross, but did not try it at this time due to the need for a completely different workflow to use Docker and concerns about slowdowns due to QEMU emulation.

Attaching these to the release would be easy as it is just one (?) additional step.

File: .github/actions/build-prqlc/action.yaml
Comment (eitsupi) [reply .github/actions/build-prqlc/action.yaml:18]: I'm sorry, but I didn't understand your intent.

I do not think it is necessary to set a prefix key here, as I believe `Swatinem/rust-cache` will detect changes of Rust source code.
Just copied from the other workflows.

**Comment by eitsupi** [comment]
Thanks to the advice I received in rust-lang/stacker#80, I am now able to build for Linux arm64.
(I have confirmed that the prqlc binary works on my Raspberry Pi 4)

image

Once this PR is merged I will work on attaching these to the releases.

File: .github/actions/build-prqlc/action.yaml
Comment (eitsupi) [reply .github/actions/build-prqlc/action.yaml:18]: Oh, I did not understand how the prefix key is automatically updated. I have update the Cargo.toml.

File: .github/workflows/pull-request.yaml
Comment (max-sixty) [reply .github/workflows/pull-request.yaml:112]: OK great, we can try it. 

I do notice our CI queue filling up occasionally, and this would almost double the number of jobs. So if it becomes an issue we could always add a `pr-build-artifacts` label and run it based on that.

Definitely fine to try, I agree that would be a nice thing to have access to.

File: .github/actions/build-prqlc/action.yaml
Comment (max-sixty) [reply .github/actions/build-prqlc/action.yaml:18]: Thanks a lot!

**Review by max-sixty — APPROVED** [review]
Looks great! 

I see there's an unrelated failure — don't let that delay merging this, I'll have a look at that later unless anyone else gets to it

**Pull request merged**
